### PR TITLE
Introduce union tag type

### PR DESCRIPTION
**Pull request by kroening**
This replaces the overloaded use of ID_symbol for union tags.

### CODE REVIEW

**Review by smowton — APPROVED** [review]

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -845,8 +846,19 @@ void goto_symex_statet::rename(
...
+    type=symbol.type;
+    rename(type, l1_identifier, ns, level);
+  }
+  else if(type.id()==ID_union_tag)
+  {
+    const symbolt &symbol=ns.lookup(to_union_tag_type(type));
+    type=symbol.type;
+    rename(type, l1_identifier, ns, level);
+  }
+  else if(type.id()==ID_struct_tag)
```
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.cpp]: That's beyond what's claimed in the commit message.

File: src/ansi-c/c_typecheck_type.cpp [hunk 2]
```diff
@@ -93,6 +93,9 @@ void c_typecheck_baset::typecheck_type(typet &type)
     typecheck_symbol_type(to_symbol_type(type));
   else if(type.id() == ID_typedef_type)
     typecheck_typedef_type(type);
+  else if(type.id() == ID_union_tag)
+  {
+  }
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: This deserves a comment that it is intentionally left blank (explaining where else the work is done, or that no work is necessary).

File: src/goto-symex/rewrite_union.cpp [hunk 3]
```diff
@@ -20,76 +20,68 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <util/c_types.h>
 
-static bool have_to_rewrite_union(
-  const exprt &expr,
-  const namespacet &ns)
+static bool have_to_rewrite_union(const exprt &expr)
```
Comment (tautschnig) [inline src/goto-symex/rewrite_union.cpp]: This commit should really be in a separate pull request. It's likely the second commit that breaks things, while this one ought to have no impact at all. Also: the commit message could be much clearer because all that happens is removing the unused namespace parameter.

**Review by tautschnig — COMMENTED** [review]
Please, let's try to move to a model where one pull request has one feature clearly described in its commit message. It will make reviewing, approving, and merging so much easier and quicker!

File: src/ansi-c/c_typecheck_type.cpp (hunk 2 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_type.cpp]: Done

**Comment by kroening** [comment]
I've turned the last commit into a separate PR (#2915).

File: src/cpp/cpp_typecheck_compound_type.cpp [hunk 4]
```diff
@@ -135,6 +135,7 @@ void cpp_typecheckt::typecheck_compound_type(
   cpp_scopet *dest_scope=nullptr;
   bool has_body=type.find(ID_body).is_not_nil();
   bool tag_only_declaration=type.get_bool(ID_C_tag_only_declaration);
+  bool is_union=type.id()==ID_union;
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_compound_type.cpp]: `const` and clang-format style spacing

File: src/ansi-c/c_typecheck_type.cpp [hunk 5]
```diff
@@ -820,12 +820,24 @@ void c_typecheck_baset::typecheck_compound_type(struct_union_typet &type)
...
   }
 
-  symbol_typet symbol_type(identifier);
-  symbol_type.add_source_location()=type.source_location();
+  if(type.id() == ID_union)
+  {
+    union_tag_typet tag_type(identifier);
+    tag_type.add_source_location() = type.source_location();
+
+    c_qualifierst original_qualifiers(type);
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: This seems common to both branches and could be done beforehand.

File: src/ansi-c/c_typecheck_type.cpp [hunk 6]
```diff
@@ -820,12 +820,24 @@ void c_typecheck_baset::typecheck_compound_type(struct_union_typet &type)
...
+    type.swap(tag_type);
+    original_qualifiers.write(type);
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: This seems common to both branches and could be done afterwards.

**Review by tautschnig — APPROVED** [review]
Many thanks for factoring out the additional changes! Looks good module the below nit-picks, and of course also CI passing.

File: src/ansi-c/c_typecheck_type.cpp (hunk 5 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_type.cpp]: done

File: src/ansi-c/c_typecheck_type.cpp (hunk 6 above)
Comment (kroening) [reply src/ansi-c/c_typecheck_type.cpp]: also done

File: src/cpp/cpp_typecheck_compound_type.cpp (hunk 4 above)
Comment (kroening) [reply src/cpp/cpp_typecheck_compound_type.cpp]: done

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 2b5277e200e437b05775e0aa2effc8bbfbc05588).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84111556
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Comment by tautschnig** [comment]
The TG failure is worrying. Otherwise it would be ready to go and unblock #2915.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7f12c59e3a7f263b792d7713c2ac5b3a93e6881e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84113136
